Created a Cookbook for Cross Validation on Linear SVM

FaroukY opened the pull request:
Created an example and a Cookbook for Cross Validation on Linear SVM Model. 

Regarding Issue #4107

karlnapf commented:
Great! Thanks for the patch. Will review tomorrow

karlnapf commented on examples/meta/src/evaluation/cross_validation_linear_svm.sg:
```diff
@@ -0,0 +1,38 @@
+CSVFile f_feats_train("../../data/classifier_binary_2d_linear_features_train.dat")
```
I would definitely also evaluate the test error, then you can say in the notebook that xvalidation can be used to estimate the test error. THis requires loading test data

FaroukY replied:
Sure, I've now included into the example evaluating the test error.

karlnapf commented on examples/meta/src/evaluation/cross_validation_linear_svm.sg:
```diff
@@ -0,0 +1,38 @@
+CSVFile f_labels_train("../../data/classifier_binary_2d_linear_labels_train.dat")
+
+#![create_features]
+RealFeatures combined_features(f_feats_train)
```
I wouldnt use combined features, this is more multiple kernel learning

FaroukY replied:
Changed them to features_train and features_test respectively.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation_linear_svm.rst:
```diff
@@ -0,0 +1,62 @@
+
+.. sgexample:: cross_validation_linear_svm.sg:evaluate_and_get_result
+
+We get the :math:`mean` of all the evaluation results and its standard deviation :math:`stddev`.
```
pls don't math mode "mean" and "stddev" but just say mean and standard deviation

FaroukY replied:
No Problem. Removed that

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation_linear_svm.rst:
```diff
@@ -0,0 +1,62 @@
+and assign that as the accuracy of this mode.
+
+
+In this tutorial, we will be doing :math:`K-` Cross Validation on a linear support vector machine (SVM) model.
```
":math:`k`-fold cross-validation"

karlnapf replied:
Pls remove "in this tutorial". Rather say "We do perform k-fold cross-validation with a linear SVM model.

FaroukY replied:
Done

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation_linear_svm.rst:
```diff
@@ -0,0 +1,62 @@
+:math:`P_1, P_2, ..., P_n` each with :math:`10` data points. Then, the model would be trained on
+:math:`P_1, P_2, ...,P_{n-1}` and test it on :math:`P_n` to get accuracy :math:`a_1` . Then the model
+would train again from scratch on :math:`P_1, P_2, ...,P_{n-2}, P_{n}` and is tested on :math:`P_{n-1}` to get
+:math:`a_2`.  This process is repeated until we have :math:`a_1, a_2, ..., a_n`, at which we take their average,
```
Pls start every sentence in a new line

FaroukY replied:
Done

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation_linear_svm.rst:
```diff
@@ -0,0 +1,62 @@
+:math:`P_1, P_2, ..., P_n` each with :math:`10` data points. Then, the model would be trained on
+:math:`P_1, P_2, ...,P_{n-1}` and test it on :math:`P_n` to get accuracy :math:`a_1` . Then the model
+would train again from scratch on :math:`P_1, P_2, ...,P_{n-2}, P_{n}` and is tested on :math:`P_{n-1}` to get
+:math:`a_2`.  This process is repeated until we have :math:`a_1, a_2, ..., a_n`, at which we take their average,
```
"This process is repeated  ... , of which we compute the average and standard deviation."

Remove: and assign that as the accuracy of this mode -- not clear what it means

FaroukY replied:
Done

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation_linear_svm.rst:
```diff
@@ -0,0 +1,62 @@
+-------
+Example
+-------
+We'll use as example a classification problem solvable by using Linear SVM or :sgclass:`CLibLinear` .
```
Simplify: "We use a linear SVM model, see ...."

FaroukY replied:
Simplified the expression

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+Cross Validation
+============================================
+
+Cross Validation is a type of model validation technique for figuring out how the
```
Could you put every sentence in a new line. This makes later patches easier to review as the diff is smaller

karlnapf replied:
"Cross validation (lower case v) is a technique to estimate an algorithm's performance on unseen data" would be a bit more concise

FaroukY replied:
Sorry for all the capitalizations (trying to get rid of that habit). I've tried to remove all capitalizations and make every sentence on a new line to reduce the diff.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+Cross Validation is a type of model validation technique for figuring out how the
+results from a statistical analysis will generalize to an independent data set. It
+essentially involves training the model on a subset of the data, and testing on the
+remainder, then repeating this process for several complementary partitions and then
```
remainder -> remaining data

FaroukY replied:
Done

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+results from a statistical analysis will generalize to an independent data set. It
+essentially involves training the model on a subset of the data, and testing on the
+remainder, then repeating this process for several complementary partitions and then
+combining all the validation results generated (For instance, an average) to estimate
```
Could you split the sentence? Lots of "ands". Also watch your grammar
(sorry need to be a bit picky here as this is user documentation)

FaroukY replied:
Sorry about the grammar. I was writing this at 2 am so I wasn't in my best form. Will make sure I keep it in mind.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+essentially involves training the model on a subset of the data, and testing on the
+remainder, then repeating this process for several complementary partitions and then
+combining all the validation results generated (For instance, an average) to estimate
+a performance for the final predictive model.
```
remove "a performance for the final predictive model"

FaroukY replied:
Done

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+
+One type of cross validation is known as :math:`k-` fold cross validation.
+
+For instance, Suppose you have the training data :math:`\mathbf{x}_1,\mathbf{x}_2, ..., \mathbf{x}_{10n}`
```
* Remove "For instance"
* replace "you" with "we"
* Please just use "n" data, not 10n

FaroukY replied:
Done

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+One type of cross validation is known as :math:`k-` fold cross validation.
+
+For instance, Suppose you have the training data :math:`\mathbf{x}_1,\mathbf{x}_2, ..., \mathbf{x}_{10n}`
+for a positive integer :math:`n` , then :math:`10-` fold cross validation might for instance partition
```
remove "might for instance"

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+
+For instance, Suppose you have the training data :math:`\mathbf{x}_1,\mathbf{x}_2, ..., \mathbf{x}_{10n}`
+for a positive integer :math:`n` , then :math:`10-` fold cross validation might for instance partition
+the data into :math:`P_1, P_2, ..., P_n` each with :math:`10` data points.
```
just say "each of equal size"

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+for a positive integer :math:`n` , then :math:`10-` fold cross validation might for instance partition
+the data into :math:`P_1, P_2, ..., P_n` each with :math:`10` data points.
+
+Next, the model would be trained on :math:`P_1, P_2, ...,P_{n-1}` and is tested on :math:`P_n` to get accuracy
```
replace acuracy with "performance measure"

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+Then the model would train again from scratch on :math:`P_1, P_2, ...,P_{n-2}, P_{n}` and is tested on :math:`P_{n-1}`
+to get :math:`a_2`.
+
+This process is repeated until we have :math:`a_1, a_2, ..., a_n`, at which we take their average and calculate
```
, of which we compute mean and standard deviation

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+This process is repeated until we have :math:`a_1, a_2, ..., a_n`, at which we take their average and calculate
+their standard deviation.
+
+We will perform :math:`k-` fold cross-validation with both a classification model (Linear SVM Model) and a
```
Don't switch times
"We perform"

karlnapf replied:
(linear SVM)

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+their standard deviation.
+
+We will perform :math:`k-` fold cross-validation with both a classification model (Linear SVM Model) and a
+regression model (Linear Ridge Regression).
```
linear ridge regression (no caitalization)

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+regression model (Linear Ridge Regression).
+
+-------
+Example of :math:`k-` Fold Cross-Validation on a Binary Classifier
```
Just "Example"

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+-------
+Example of :math:`k-` Fold Cross-Validation on a Binary Classifier
+-------
+We will use Linear SVM Model. (see :doc:`../binary/linear_support_vector_machine` for a more complete example of Linear SVM usage).
```
punctuation problems. Also same comments as above regarding capitalization

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+.. sgexample:: cross_validation.sg:create_features
+
+Next,we initialize a splitting strategy :sgclass:`CStratifiedCrossValidationSplitting`, which is needed
+to divide the dataset into :math:`k-` folds for the :math:`k-` fold cross validation (In this case, we make :math:`k=2` ).
```
remove the "in this case", this is obvious from the code and is easier to maintain in case the code changes

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+Next,we initialize a splitting strategy :sgclass:`CStratifiedCrossValidationSplitting`, which is needed
+to divide the dataset into :math:`k-` folds for the :math:`k-` fold cross validation (In this case, we make :math:`k=2` ).
+
+We also have to decide on an evaluation criterion class (From :sgclass:`CEvaluation`) to evaluate the performance of the trained models. For this binary classifier,
```
from

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+to divide the dataset into :math:`k-` folds for the :math:`k-` fold cross validation (In this case, we make :math:`k=2` ).
+
+We also have to decide on an evaluation criterion class (From :sgclass:`CEvaluation`) to evaluate the performance of the trained models. For this binary classifier,
+we use :sgclass:`CAccuracyMeasure` to evaluate the accuracy of the model.
```
the second sentence repeats most of the first sentence, pls merge them

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+
+.. sgexample:: cross_validation.sg:create_cross_validation
+
+Finally, we evaluate the model and get the results (a :sgclass:`CCrossValidationResult` instance).
```
Repeated use of "finally"

FaroukY replied:
Removed it and made the flow slightly better.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,101 @@
+
+.. sgexample:: cross_validation.sg:evaluate_and_get_result
+
+We get the mean of all the evaluation results and their standard deviation stddev.
```
remove stddev
remove "the" in "the evaluation results"

FaroukY replied:
Done

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation_multiple_kernel_learning_weights_storage.rst:
```diff
@@ -2,11 +2,9 @@
-the performance of the final predictive model.
+For more information on Cross Validation, check :doc:`./cross_validation`.
+
+We will show an example of cross validation on Multiple Kernel Learning.
```
We now show

FaroukY replied:
Added it.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation_multiple_kernel_learning_weights_storage.rst:
```diff
@@ -2,11 +2,9 @@
-it is based on training and test the model many times on different complementary partitions of the original
-training dataset and then to combine the validation results (e.g. average) to estimate a
-the performance of the final predictive model.
+For more information on Cross Validation, check :doc:`./cross_validation`.
```
no capitals, 
The notebook should have an intro sentence: something like "In this example, we illustrate how to use cross-validation with multiple kernel learning (MKL)."

FaroukY replied:
Added the sentence to the top of the file.

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,111 @@
+
+#
```
pls remove these empty comment lines

vigsterkr commented:
@FaroukY as the cookbooks are actually automatically used as integration tests y, ou should add the generated integration tests files to the https://github.com/shogun-toolbox/shogun-data repository.
see details here: https://github.com/shogun-toolbox/shogun/blob/develop/doc/readme/DEVELOPING.md#integration-testing-of-results

karlnapf commented:
Any updates here?

FaroukY commented:
@karlnapf  Working on it today. Was trying to get Stan math to work with Shogun.

FaroukY commented:
@karlnapf @vigsterkr I am a bit confused on the submodule. The submodule is not owned by me, so when I push it, it denies me. I tried committing to the data submodule then committing that change in the shogun folder but it made the last commit which doesn't seem right. Any ideas? Right now, my data submodule branch is 1 commit (The integration test) ahead of origin/master. How would I pull request there?

vigsterkr commented:
@FaroukY regarding the data submodule you have to do the same thing as with the shogun repo. basically fork it prepare a patch and create a pull request

FaroukY commented:
@vigsterkr  Got it. I sent the pull request, and undid my last commit

vigsterkr commented:
try to rebase now and tie the data submodule to the HEAD of master branch... that should make the unit tests pass

FaroukY commented:
@vigsterkr  This is strange. I just rebased with shogun-data and the Travis build failed. I looked at why it failed and apparently, it failed the cross_validation integration test that I introduced, and ideas where this might've gone wrong?

FaroukY commented:
@vigsterkr  I think I know what the problem is but I am not sure how to solve it. StratifiedCrossValidationSplitting is used in this example which "randomly" creates a cross-validation splitting of the data. So the testing and training accuracy every time are changed when you run the tests (Depending on the cross-validation splits) so each interface (cpp, python, ...) generates a different .dat file (but with very close values). For example, the accuracy variable in one might be 0.92 while in another 0.916 since they used different partitions to train and test. Is there a workaround for this?

karlnapf commented:
Hi there. Did you try fixing the seed?
Alternatively, you could just not store the accuracy in a variable (all variables are used in integration tests, so if you dont store a number, it wont be used)

karlnapf commented:
Check the other cookbooks that depend on random states and how they fix the seed

vigsterkr commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,108 @@
+CSVFile f_labels_train("../../data/classifier_binary_2d_linear_labels_train.dat")
+CSVFile f_labels_test("../../data/classifier_binary_2d_linear_labels_test.dat")
+
+
```
set the seed of random here, with:
```Math:init_random(1)```

vigsterkr commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,108 @@
+CSVFile reg_feats_test("../../data/regression_1d_linear_features_test.dat")
+CSVFile reg_lab_train("../../data/regression_1d_linear_labels_train.dat")
+CSVFile reg_lab_test("../../data/regression_1d_linear_labels_test.dat")
+
```
you probably wanna reset the seed of the prng:
```Math:init_random(1)```

FaroukY commented:
@vigsterkr  This is weird. The integration test is passing on my local setup (Generating the same .dat file) but failing on CI even though I updated shogun-data with the correct .dat integration test. Any idea?

FaroukY commented:
@vigsterkr  Sorry for the inconvenience. The .dat file I uploaded last time was the .dat on a different branch, and so had the wrong values again :/ . Sorry for the trouble, this **should** be the last pull request hopefully in **shogun-data** and this pr should pass the tests. I've verified twice it compiles using Docker.

karlnapf commented:
CI is still failing. I merged the data PR in the meantime

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,113 @@
+CSVFile f_labels_test("../../data/classifier_binary_2d_linear_labels_test.dat")
+Math:init_random(2)
+
+
```
why all the whitespace?

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,113 @@
+#![evaluate_and_get_result]
+
+#![get_results]
+real cv_mean_accuracy = result.get_mean()
```
could you try:
```
result = cross.evaluate()
real cv_mean_accuracy = result.get_real("mean")
```
Might have to check under which name the mean is registered

karlnapf replied:
you now need to do this as we dont do the obtain_from_generic anymore

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,113 @@
+#![get_results_test_data]
+
+
+
```
too much whitespace

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,113 @@
+#  Cross Validation for Binary Classifier
```
for a binary classifier
or
for binary classification

karlnapf replied:
This is still unaddressed

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+============================================
```
the === should fit the heading

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+Cross Validation
+============================================
+
+Cross validation is a type of model validation technique to estimate an algorithm's performance on unseen data.
```
remove "is a type of model validation technique" and replace with
"allows to"

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+
+Cross validation is a type of model validation technique to estimate an algorithm's performance on unseen data.
+
+It essentially involves training the model on a subset of the data, and testing on the
```
replace "essentially" with "is based on"

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+remaining data.
+
+The process is repeated for several complementary partitions and
+combines all the validation results generated (For instance, an average).
```
grammar error make it
"and all validation results are combined, e.g. by averaging"

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+The process is repeated for several complementary partitions and
+combines all the validation results generated (For instance, an average).
+
+One type of cross validation is known as :math:`k-` fold cross validation.
```
remove "known as"

karlnapf replied:
remove whitespace after "k-"

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+One type of cross validation is known as :math:`k-` fold cross validation.
+
+Suppose we have the training data :math:`\mathbf{x}_1,\mathbf{x}_2, ..., \mathbf{x}_{N}`
+for a positive integer :math:`n` , then :math:`k-` fold cross validation partitions
```
inconsistent N and n,
no whitespace after n pls

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+
+Suppose we have the training data :math:`\mathbf{x}_1,\mathbf{x}_2, ..., \mathbf{x}_{N}`
+for a positive integer :math:`n` , then :math:`k-` fold cross validation partitions
+the data into :math:`P_1, P_2, ..., P_n` each with equal size (Where :math:`n=\frac{N}{k}`).
```
say "with approximately equal size"
and remove the (where n=...)

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+
+Next, the model would be trained on :math:`P_1, P_2, ...,P_{n-1}` and is tested on :math:`P_n` to get performance measure
+:math:`a_1` .
+
```
remove whitespace, make this one paragraph

karlnapf replied:
Dont use tenses like "would be trained", just say "is trained" (also in the other places)

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+the data into :math:`P_1, P_2, ..., P_n` each with equal size (Where :math:`n=\frac{N}{k}`).
+
+Next, the model would be trained on :math:`P_1, P_2, ...,P_{n-1}` and is tested on :math:`P_n` to get performance measure
+:math:`a_1` .
```
why a newline?

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+Then the model would train again from scratch on :math:`P_1, P_2, ...,P_{n-2}, P_{n}` and is tested on :math:`P_{n-1}`
+to get :math:`a_2`.
+
+This process is repeated until we have :math:`a_1, a_2, ..., a_n`, at which we can compute their mean and standard deviation.
```
more grammar errors in here
replace "at which we can" with
"and then we can compute"

karlnapf replied:
Also you are between direct and indirect formulations.

"the model is trained" (indirect) and
"we train the model"

pls be consistent within this cookbook

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+We perform :math:`k-` fold cross-validation with both a classification model (linear SVM) and a
+regression model (linear ridge regression).
+
+-------
```
pls match the heading

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+Example of a binary Classifier
+-------
+
+We use a linear SVM model. (see :doc:`../binary/linear_support_vector_machine` for a more complete example of linear SVM usage).
```
Replace period with coma and remove parenthesis

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+.. sgexample:: cross_validation.sg:create_features
+
+Next,we initialize a splitting strategy :sgclass:`CStratifiedCrossValidationSplitting`, which is needed
+to divide the dataset into :math:`k-` folds for the :math:`k-` fold cross validation.
```
so space after k-

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+
+We also have to decide on an evaluation criterion class (from :sgclass:`CEvaluation`) to evaluate the performance of the trained models.
+
+In this case, we use :sgclass:`CAccuracyMeasure`
```
missing period

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+We also have to decide on an evaluation criterion class (from :sgclass:`CEvaluation`) to evaluate the performance of the trained models.
+
+In this case, we use :sgclass:`CAccuracyMeasure`
+
```
pls dont make paragraphs with single sentences

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,102 @@
+
+
+-------
+Example of regression
```
for this section, I have the same comments as above

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,113 @@
+#  Cross Validation for Binary Classifier
+
+
+CSVFile f_feats_train("../../data/classifier_binary_2d_linear_features_train.dat")
```
could you use the new API (see other examples)

FaroukY commented:
@karlnapf Improved the grammar and addressed the comments. I also updated the example to the new API. However, that meant changing some stuff in the .sg file, and so the generated .dat file is different. I'll wait until you give me a green light that the .sg file is okay before raising a pull request on shogun-data so we don't have to do it several times.

karlnapf commented:
Sure about the data.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+
+It is based on training the model on a subset of the data, and testing on the
+remaining data.
+
```
minor: no paragraph break, combine with previous

FaroukY replied:
Sure, Fixed on the next commit

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+results are combined, e.g. by averaging.
+
+One type of cross validation is :math:`k-`fold cross validation.
+
```
no paragraph break. Colon after sentence

FaroukY replied:
Fixed in the next commit

karlnapf replied:
Pls only comment if you have questions or comments. No need to say that you are changing this in the next commit, just send the commit :) (trying to keep the noise in my inbox down)

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+
+Suppose we have the training data :math:`\mathbf{x}_1,\mathbf{x}_2, ..., \mathbf{x}_{N}`
+for a positive integer :math:`N`, then :math:`k-`fold cross validation partitions
+the data into :math:`P_1, P_2, ..., P_n` each with approximately equal size. Next, we train the model on
```
pls start new sentences in a new line, makes later editing much easier

FaroukY replied:
Sure, will be fixed in the next commit. 

So by this, do you mean:

Sentence 1 .....
Sentence 2 .....

or 

Sentence 1....

Sentence 2....

?

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+One type of cross validation is :math:`k-`fold cross validation.
+
+Suppose we have the training data :math:`\mathbf{x}_1,\mathbf{x}_2, ..., \mathbf{x}_{N}`
+for a positive integer :math:`N`, then :math:`k-`fold cross validation partitions
```
period instead of coma after "N", and remove "then"

iglesias replied:
It looks like there is a verb missing here, probably divide.

FaroukY replied:
@karlnapf  Fixed in next commit.

@iglesias Wouldn't "partitions" work here as a verb?

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+for a positive integer :math:`N`, then :math:`k-`fold cross validation partitions
+the data into :math:`P_1, P_2, ..., P_n` each with approximately equal size. Next, we train the model on
+:math:`P_1, P_2, ...,P_{n-1}` and test it on :math:`P_n` to get performance measure
+:math:`a_1`. Finally we train the model again from scratch on :math:`P_1, P_2, ...,P_{n-2}, P_{n}` and test it on :math:`P_{n-1}`
```
I would remove the last examplek as you already said twice what is happening ("Finally we train the model again from scratch on :math:`P_1, P_2, ...,P_{n-2}, P_{n}` and test it on :math:`P_{n-1}`")

FaroukY replied:
@karlnapf  In the last sentence, I use ``a_1, a_2, ..., a_n`` so I think leaving this sentence would be important so as to avoid confusion at where ``a_2, ..., a_n`` came from?

karlnapf replied:
I think this is clear from the first two examples. If you disagree we can also leave it in.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+the data into :math:`P_1, P_2, ..., P_n` each with approximately equal size. Next, we train the model on
+:math:`P_1, P_2, ...,P_{n-1}` and test it on :math:`P_n` to get performance measure
+:math:`a_1`. Finally we train the model again from scratch on :math:`P_1, P_2, ...,P_{n-2}, P_{n}` and test it on :math:`P_{n-1}`
+to get :math:`a_2`. This process is repeated until we have :math:`a_1, a_2, ..., a_n`, and then we
```
pls newline after each sentence end

FaroukY replied:
Fixed in new commit.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+regression model (linear ridge regression).
+
+------------------------------
+Example of a binary Classifier
```
"Example for a binary classifier"
(no capital B, of->for)

FaroukY replied:
Fixed in next commit.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+.. sgexample:: cross_validation.sg:create_features
+
+Next,we initialize a splitting strategy :sgclass:`CStratifiedCrossValidationSplitting`, which is needed
+to divide the dataset into :math:`k-`folds for the :math:`k-` fold cross validation. We also have to decide on an evaluation
```
one sentence per line, no linearbreaks within sentences (I wont comment again, this is globally :) )

FaroukY replied:
Fixed in next commit.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+
+
+---------------------
+Example of regression
```
Example for regression

FaroukY replied:
Fixed in next commit.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+
+.. sgexample:: cross_validation.sg:create_features_REGRESSION
+
+Next, we initialize a splitting strategy :sgclass:`CCrossValidationSplitting` (Do not use :sgclass:`CStratifiedCrossValidationSplitting`
```
Do not use -> "Stratified cross-validation only makes sense for classification problems"

FaroukY replied:
Changed in next commit

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+
+Next, we initialize a splitting strategy :sgclass:`CCrossValidationSplitting` (Do not use :sgclass:`CStratifiedCrossValidationSplitting`
+with regression), which is needed to divide the dataset into :math:`k-` folds for the :math:`k-` fold cross validation. We also have to decide on an
+evaluation criterion class (from :sgclass:`CEvaluation`) to evaluate the performance of the trained models. Here, we use :sgclass:`CMeanSquaredError`.
```
pls dont copy paste, many of those things were said above already

karlnapf replied:
I.e. you can be briefer in the second example

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,94 @@
+#  Cross Validation for Binary Classifier
+CSVFile f_feats_train("../../data/classifier_binary_2d_linear_features_train.dat")
+CSVFile f_feats_test("../../data/classifier_binary_2d_linear_features_test.dat")
+CSVFile f_labels_train("../../data/classifier_binary_2d_linear_labels_train.dat")
```
pls use the new API everywhere

`File f_feats_train 0 csv_file(....`

FaroukY replied:
Fixed in coming commit

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,94 @@
+
+#![create_features]
+Features features_train = features(f_feats_train)
+BinaryLabels labels_train(f_labels_train)
```
use Labels and the new API

FaroukY replied:
Now uses Labels. Fixed in the coming commit.

FaroukY replied:
Tried changing to Labels but has problems with StratifiedCrossValidation (I opened an issue with details) so I'm leaving it BinaryLabels for now.

karlnapf replied:
As said in #4299 you will need to patch the .cpp file and convert the labels to the right type rather than asserting it.

iglesias replied:
Did you attempt doing this, @FaroukY?

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,94 @@
+#OK
+#![create_instance]
+Machine svm = machine("LibLinear", labels=labels_train)
+svm.put("liblinear_solver_type", enum LIBLINEAR_SOLVER_TYPE.L2R_L2LOSS_SVC)
```
pls use kwargs in the meta example, not put

karlnapf replied:
we don't use `put` in meta examples, you would do
`Machine svm = machine("LibLinear", labels=labels_train, liblinear_solver_type = enum LIBLINEAR_SOLVER_TYPE.L2R_L2LOSS_SVC)`

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,94 @@
+#![create_instance]
+
+#![create_cross_validation]
+StratifiedCrossValidationSplitting splitting_strategy(labels_train, 2)
```
ah we need to add factories for this ... it is OK to use the old API here for now

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,94 @@
+
+#![create_cross_validation]
+StratifiedCrossValidationSplitting splitting_strategy(labels_train, 2)
+AccuracyMeasure evaluation_criterion()
```
new API and factory pls, see other examples

FaroukY replied:
@karlnapf  Other examples in cross_validation use AccuracyMeasure evaluation_criterion(), do you mean another example?

karlnapf replied:
`evaluation_cirterion = evaluation("AccuracyMeasure")`

iglesias replied:
What about this one, @FaroukY?

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,94 @@
+#![create_cross_validation]
+
+#![evaluate_and_get_result]
+CrossValidationResult result = CrossValidationResult:obtain_from_generic(cross.evaluate())
```
pls no obtain_from_generic, you dont need it, as you are using `get` downstream

FaroukY replied:
Fixed in next commit.

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,94 @@
+
+#![get_results_test_data]
+svm.train(features_train)
+BinaryLabels labels_predict = svm.apply_binary(features_test)
```
pls use `apply` and `Labels` (new API)

FaroukY replied:
Fixed in next commit.

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,94 @@
+CSVFile reg_feats_train("../../data/regression_1d_linear_features_train.dat")
+CSVFile reg_feats_test("../../data/regression_1d_linear_features_test.dat")
+CSVFile reg_lab_train("../../data/regression_1d_linear_labels_train.dat")
+CSVFile reg_lab_test("../../data/regression_1d_linear_labels_test.dat")
```
for this example, same comments as for the above

FaroukY replied:
Done in next commit.

iglesias commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+Cross validation allows to estimate an algorithm's performance on unseen data.
+
+It is based on training the model on a subset of the data, and testing on the
+remaining data.
```
I propose being more precise here. All the available data is divided into train and test subsets. Then it is in fact the train subset which is further divided to perform cross-validation. That is, you train using part of the train subset and validate using another part of the train subset (we could say, the "validation subset"). This is of course still simplified, but it does make clear that the data used for validation is part of the train data. And more importantly, it is also clear that the test subset is never used for any training.

iglesias commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+It is based on training the model on a subset of the data, and testing on the
+remaining data.
+
+The process is repeated for several complementary partitions and and all validation
```
Typo: and and.

iglesias commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+to get :math:`a_2`. This process is repeated until we have :math:`a_1, a_2, ..., a_n`, and then we
+can compute their mean and standard deviation.
+
+We will perform :math:`k-` fold cross-validation with both a classification model (linear SVM) and a
```
It would be nice to keep the hyphenation in cross-validation consistent.

iglesias commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+Next,we initialize a splitting strategy :sgclass:`CStratifiedCrossValidationSplitting`, which is needed
+to divide the dataset into :math:`k-`folds for the :math:`k-` fold cross validation. We also have to decide on an evaluation
+criterion class (from :sgclass:`CEvaluation`) to evaluate the performance of the trained models.
+In this case, we use :sgclass:`CAccuracyMeasure`. We then instantiate a :sgclass:`CCrossValidation` instance and set the number of cross validation's runs.
```
Redundant: instantiate an instance.

iglesias commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,90 @@
+Next, we initialize a splitting strategy :sgclass:`CCrossValidationSplitting` (Do not use :sgclass:`CStratifiedCrossValidationSplitting`
+with regression), which is needed to divide the dataset into :math:`k-` folds for the :math:`k-` fold cross validation. We also have to decide on an
+evaluation criterion class (from :sgclass:`CEvaluation`) to evaluate the performance of the trained models. Here, we use :sgclass:`CMeanSquaredError`.
+We then instantiate a :sgclass:`CCrossValidation` instance and set the number of cross validation's runs.
```
Same as above regarding instance redundancy. Please check and fix if there are more instances (lol) of this.

karlnapf commented:
What is the state of this?
@FaroukY It would be cool if you could address suggestions asap, otherwise the whole thing gets stale and is more work to pick up again.

FaroukY commented:
Alright, Next round :).  Addressed most of your points (Except one point which I commented on).

karlnapf commented:
https://travis-ci.org/shogun-toolbox/shogun/jobs/382825416#L4800

Integration test fails. Make sure to reproduce it locally and potentially update the data file

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,81 @@
+Finally we train the model again from scratch on :math:`P_1, P_2, ...,P_{n-2}, P_{n}` and test it on :math:`P_{n-1}` to get :math:`a_2`.
+This process is repeated until we have :math:`a_1, a_2, ..., a_n`, and then we can compute their mean and standard deviation.
+
+We will perform :math:`k-` fold cross validation with both a classification model (linear SVM) and a regression model (linear ridge regression).
```
link to the classes

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,81 @@
+
+.. sgexample:: cross_validation.sg:create_features
+
+Next,we initialize a splitting strategy :sgclass:`CStratifiedCrossValidationSplitting`, which is needed to divide the dataset into :math:`k-`folds for the :math:`k-` fold cross validation.
```
missing whitespace

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,81 @@
+.. sgexample:: cross_validation.sg:create_features_REGRESSION
+
+Next, we initialize a splitting strategy :sgclass:`CCrossValidationSplitting` (:sgclass:`CStratifiedCrossValidationSplitting` only makes sense for classification problems), to divide the dataset into :math:`k-` folds for the :math:`k-` fold cross validation.
+We also have to decide on an evaluation criterion class (from :sgclass:`CEvaluation`) to evaluate the performance of the trained models.
```
as said before, pls dont copy paste sentences from above. This gets to verbose. Either adjust the sentences "Here, we use CMeanSquaredError as an evaluation metric", or just remove this at all.

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,81 @@
+
+.. sgexample:: cross_validation.sg:create_cross_validation_REGRESSION
+
+Finally, we evaluate the model and get the results (a :sgclass:`CCrossValidationResult` instance).
```
Same here, pls dont copy paste the text. I think most of this can be removed, and you can just group all the snippets in one and just say something like "Here is an illustration how to do this with a regression algorithm"

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation_multiple_kernel_learning_weights_storage.rst:
```diff
@@ -1,12 +1,9 @@
-it is based on training and test the model many times on different complementary partitions of the original
-training dataset and then to combine the validation results (e.g. average) to estimate a
-the performance of the final predictive model.
+For more information on cross-validation, check :doc:`./cross_validation`.
```
+1!

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,91 @@
+#Cross Validation for Binary Classifier
```
not sure this comment is relevant?

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,91 @@
+#![create_instance]
+
+#![create_cross_validation]
+StratifiedCrossValidationSplitting splitting_strategy(labels_train, 2)
```
could you patch `factory.h` to have the cross-validation splitting as part of it? We wanna use the new API everywhere if possible

iglesias replied:
This is a good one indeed, but for the sake of going ahead with this PR, I suggest to do that independently. Are we keeping track of the stuff that is still missing factories somewhere?

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,91 @@
+crossReg.set_num_runs(2)
+#![create_cross_validation_REGRESSION]
+
+#![evaluate_and_get_result_REGRESSION]
```
as said, I would block most of this into a single snippet and then simplify the example above

karlnapf commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,91 @@
+#![create_cross_validation_REGRESSION]
+MeanSquaredError MSE_evaluation()
+CrossValidationSplitting split_strategy(reg_labels_train, 2)
+CrossValidation crossReg(lrr, reg_features_train, reg_labels_train, split_strategy, MSE_evaluation)
```
pls dont mix camelCase with camel_case.

iglesias replied:
@FaroukY `cross_validation`, why Reg?

FaroukY commented:
@karlnapf Integration tests fail because we haven't updated shogun-data with the new .dat file from the new .sg files. I am just waiting until this is a green light then I'll send a correct PR to shogun-data with the right .dat.

iglesias commented:
@FaroukY, the mentioned `StratifiedCrossValidation` is `StratifiedCrossValidationSplitting`, I guess.

I think there are other points still open. I will comment in there for convenience.

iglesias commented:
@FaroukY, I folllowed-up in a few unadressed comments from Heiko. I am confused because apart from the `StratifiedCrossValidationSplitting`, all the other points were expected to be addressed. Am I looking at something wrong?

FaroukY commented:
@iglesias Sorry, I should've been more clear. All the other comments are addressed (locally) but not yet pushed (I've addressed all the points you commented on, just haven't pushed them). There is one problem though in StratifiedCrossValidationSplitting. Using the new API (Labels l = labels(...)) leads to throwing an error. I've talked to @vigsterkr on the irc and he said he'll first need to discuss a design problem with Heiko before I change StratifiedCrossValidationSplitting to use the new API.

So in summary, all points have been addressed (locally) and I was just wondering whether I should push them or wait for the StratifiedCrossValidationSplitting to be fixed first then push.

iglesias commented:
No problem, @FaroukY. Understood.

Regarding the problem with `StratifiedCrossValidationSplitting`. Can it be solved with what Heiko mentioned in an earlier comment: patching the .cpp file and converting the labels to the right type rather than asserting?

karlnapf commented:
always push things up, dont hold them back.
The labels problem should not block anymore now #4313

FaroukY commented:
Alright, pushed. If this is okay, I'll send a pr to shogun-data with the correct .dat file.

vigsterkr commented on examples/meta/src/evaluation/cross_validation.sg:
```diff
@@ -0,0 +1,77 @@
+#![create_cross_validation]
+
+#![evaluate_and_get_result]
+CrossValidationResult result = cross.evaluate()
```
here you need to make sure that you convert it with obtain_from_generic :(

karlnapf replied:
No need for the obtain_from_generic. Just dont use the specialized type
The method signature is `virtual CEvaluationResult* evaluate();`
so you can do
`EvaluationResult result = cross.evaluate()` and then you use `get` downstream.
With any of those changes **pls try them before you send a PR** as this back and forth is quite time consuming...

karlnapf commented:
and yes, pls always send a PR with the data file, otherwise we waste time looking at error messages

FaroukY commented:
Changed obtain_from_generic ==> EvaluationResult result = cross.evaluate() and send a pr to shogun-data to fix the integration test failing

vigsterkr commented:
@FaroukY you should reference the new data commit so that actually things are being checked against the right integration dat file. note i've merged your PR into the master of shogun-data

iglesias commented:
@FaroukY, what is the state on this one?

vigsterkr commented:
i wonder why does octave fail in this case :(

vigsterkr commented:
only happens in case of octave:

```
[DEBUG] Own parameter Serializable::value=23.5918 different from provided Serializable::value=7.89437
```

karlnapf commented:
This should be fixed in develop, try rebasing and octave will pass

vigsterkr commented:
@FaroukY ? this is taking unreasonably long to get it merged

FaroukY commented:
@vigsterkr Sorry, didn't know this was fixed. Discussed it with Heiko on IRC but didn't know it got fixed. Rebased and running the tests now.

FaroukY commented:
@vigsterkr Tests pass except one of them which timed out. Restart it?

karlnapf commented on doc/cookbook/source/examples/evaluation/cross_validation.rst:
```diff
@@ -0,0 +1,66 @@
+
+.. sgexample:: cross_validation.sg:create_cross_validation_REGRESSION
+
+You can evaluate the error of the model and cross validation similar to how it was done in the case of binary .
```
minor whitespace error at the end
